rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -278,37 +273,16 @@ private void triggerBuilder(TextDocumentIdentifier doc) {
     private TextDocumentState updateContents(VersionedTextDocumentIdentifier doc, String newContents) {
         TextDocumentState file = getFile(doc);
         logger.trace("New contents for {}", doc);
-        handleParsingErrors(file, file.update(doc.getVersion(), newContents));
+        file.update(doc.getVersion(), newContents);
+        handleParsingErrors(file, file.getCurrentDiagnosticsAsync(Duration.ofMillis(800))); // Warning: Might be a later version (when a concurrent update happened)
         return file;
     }
 
-    private void handleParsingErrors(TextDocumentState file, CompletableFuture<Versioned<ITree>> futureTree) {
-        var version = file.getCurrentContent().version();
-        futureTree.handle((tree, excp) -> {
-            Diagnostic newParseError = null;
-            if (excp instanceof CompletionException) {
-                excp = excp.getCause();
-            }
-
-            if (excp instanceof Throw) {
-                Throw thrown = (Throw) excp;
-                newParseError = Diagnostics.translateRascalParseError(thrown.getException(), columns);
-            }
-            else if (excp instanceof ParseError) {
-                newParseError = Diagnostics.translateDiagnostic((ParseError)excp, columns);
-            }
-            else if (excp != null) {
-                logger.error("Parsing crashed", excp);
-                newParseError = new Diagnostic(
-                    new Range(new Position(0,0), new Position(0,1)),
-                    "Parsing failed: " + excp.getMessage(),
-                    DiagnosticSeverity.Error,
-                    "Rascal Parser");
-            }
-            logger.trace("Finished parsing tree, reporting new parse error: {} for: {}", newParseError, file.getLocation());
-            facts(file.getLocation()).reportParseErrors(file.getLocation(), version,
-                newParseError == null ? Collections.emptyList() : Collections.singletonList(newParseError));
-            return null;
+    private void handleParsingErrors(TextDocumentState file, CompletableFuture<Versioned<List<Diagnostic>>> diagnosticsAsync) {
```

Comment:
Observation: in this codebase we've tried to keep the LSP mapping in this class, and out of the other classes. For this function this has shifted to the `TextDocumentState` class. 

I'm not sure if maybe we should pass that mapping as an argument or something?